### ISSUE TITLE
Empty view snapshots

### ISSUE THREAD

**Issue by ashfurrow** [facebook/ios-snapshot-test-case#52]
Hey all. I'm having a really weird, intermittent problem when I try and snapshot a view belonging to a controller that's been loaded from a Storyboard in iOS 8. Lots of details of what I've tried so far are [here](http://stackoverflow.com/questions/26156660/empty-uiview-snapshots-from-view-controller-loaded-from-storyboard-in-xctest-tes), but the gist of it is that the image rendered from `drawViewHierarchyInRect` is empty – the correct size, but transparent. 

Has anyone else seen this?

**Comment by toddlee** [comment]
For what it's worth..

It seems like when auto layout is involved, it almost always fails. However, pausing by putting a breakpoint makes it pass which made me think this is related to timing issue.
And.. with this code,

```
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        UBSnapshotVerifyView(view, @"Should pass!");
    });
```
I confirmed that it always passes (at least I didn't get any intermittent failure).. so it is definitely related to some timing issue. Until someone has a correct fix, I'll stick to this hack..

**Comment by ashfurrow** [comment]
@toddlee That's odd – we've been using this on [our project](https://github.com/artsy/eidolon) for months with no problems. If it's not too much trouble, could you provide a sample project demonstrating the problem?

**Comment by toddlee** [comment]
@ashfurrow Sorry.. it was actually not doing its job correctly. Since that block is run after the test ends, it doesn't actually compare images.
I am still getting empty snapshots and using renderAsLayer doesn't fix it either (it renders something, but not in the correct way)

**Comment by grantjk** [comment]
@toddlee We've also had troubles with the renderAsLayer approach - views not properly being laid out  by the time the screenshot is taken. We've had to resort to running all our snapshots against the iOS7 simulator which is 100% reliable.

In our case I think the issue relates to views with nested custom views with their own constraints - or views that have conditionally updated constraints - but I haven't been able to replicate it yet in a simple project.

**Comment by toddlee** [comment]
@grantjk We have the same nested custom views with own constraints. We also have other views that don't use AutoLayout and are not seeing this issue. We're in the middle of migrating to autolayout and those views using autolayout seem to fail.
Interestingly, if only the failing test is run, it passes. However if I run the whole test suites, it fails.

**Comment by dstnbrkr** [comment]
Anyone willing to contribute a failing test case?

**Comment by grantjk** [comment]
I will see if I can whip something together in the next few days

**Comment by FGoessler** [comment]
We have some similar issues and also suspect a timing issue. Our working workaround is adding the following lines to our basic XCTestCase subclass from which all our other test cases inherit.

```objective-c
static dispatch_once_t onceToken;
dispatch_once(&onceToken, ^{
    NSDate *runUntil = [NSDate dateWithTimeIntervalSinceNow:1.0];
    [[NSRunLoop currentRunLoop] runUntilDate:runUntil];
});
```

We assume that some UIKit framework stuff isn't initialised when the tests start and waiting a little while gives the frameworks enough time to settle.

**Comment by jgapps** [comment]
@dstnbrkr Here's an example showing renderAsLayer = YES and NO producing a much different layout.  We can't use renderAsLayer = NO in iOS8 because of the aforementioned empty views. We've resorted to just recording and running everything on iOS7

https://github.com/grantjk/SnapshotLayerFailingTestCase

Hope that helps

**Comment by dstnbrkr** [comment]
this is great. thanks!

**Comment by toddlee** [comment]
@jgapps That renderAsLayer=NO/iOS8 looks pretty similar to our case. Thanks for providing a reproducible test! Looking forward to the fix.

**Comment by dstnbrkr** [comment]
If I'm understanding right, @ashfurrow describes this case:
fails when renderAsLayer = NO (i.e. drawViewHierarchyInRect)
succeeds when renderAsLayer = YES

however, @grantjk's test case demonstrates the opposite:
succeeds when renderAsLayer = NO
fails when renderAsLayer = YES

**Comment by ashfurrow** [comment]
Is it possible that we're seeing more than one problem? In [my use](https://github.com/ashfurrow/Nimble-Snapshots/blob/6309073dba0b1411a590acc6d9f1128cc5b0222d/HaveValidSnapshot.swift#L12-L18), I always call `beginAppearanceTransition()` and `endAppearanceTransition()` on the view controller before snapshotting it to avoid incorrect states. Since @grantjk's example loads the view from a nib instead of a view controller (in a nib or otherwise), I don't know what an equivalent fix would be.

**Comment by toddlee** [comment]
@ashfurrow In our test case, we instantiated a view directly without the view controller. Looks like I can't try out that `beginAppearanceTransition()` `endAppearanceTransition()`. Any suggestion?

**Comment by ashfurrow** [comment]
@toddlee I wish I did – I've never had to snapshot views that are loaded from a nib. Usually they're programmatically constructed or belong to a view controller in a storyboard/xib.

**Comment by toddlee** [comment]
@ashfurrow We construct all views programmatically too..
I'll try to create a test that reproduces our issue.

**Comment by dstnbrkr** [comment]
Tried @grantjk's [test case](https://github.com/grantjk/SnapshotLayerFailingTestCase). It doesn't show us a snapshot bug. It does illustrate the layout differences between with and without calling drawViewHierarchyInRect:afterScreenUpdates:.

I used [chisel visualize](https://github.com/facebook/chisel) to see that [self.emptyView](https://github.com/grantjk/SnapshotLayerFailingTestCase/blob/31f0d0ed21027f18335bcc63a58bf1a19695c05a/SnapshotLayerFailingTestCaseTests/ActionableEmptyViewSnapshotTests.m#L30) looks [off center](https://raw.githubusercontent.com/grantjk/SnapshotLayerFailingTestCase/master/SnapshotLayerFailingTestCaseTests/ReferenceImages/ActionableEmptyViewSnapshotTests/RenderAsLayerYES.png) before FBSnapshotVerifyView is called. Calling drawViewHierarchyInRect:afterScreenUpdates: before FBSnapshotVerifyView fixes self.emptyView's layout regardless of the value of renderAsLayer.

My guess is that drawViewHierarchyInRect:afterScreenUpdates: places self.emptyView in a containing view which has some default constraints that affect self.emptyView's layout and correct it. Maybe some tweaks to self.emptyView's constraints would fix this. For renderAsLayer to work, self.emptyView would need to be laid out correctly before FBSnapshotVerifyView is called.
